astroid/constraint.py

```diff
@@ -0,0 +1,142 @@
+# For details: https://github.com/PyCQA/astroid/blob/main/LICENSE
+# Copyright (c) https://github.com/PyCQA/astroid/blob/main/CONTRIBUTORS.txt
+
+"""Classes representing different types of constraints on inference values."""
+
+from abc import abstractmethod
+from typing import Dict, Optional, Type, TypeVar, Union
+
+from astroid import nodes, util
+
+NameNodes = Union[nodes.AssignAttr, nodes.Attribute, nodes.AssignName, nodes.Name]
+ConstraintT = TypeVar("ConstraintT", bound="Constraint")
+
+
+class Constraint:
+    """Represents a single constraint on a variable."""
+
+    def __init__(self, node: nodes.NodeNG, negate: bool) -> None:
+        self.node = node
+        """The node that this constraint applies to."""
+        self.negate = negate
+        """True if this constraint is negated. E.g., "is not" instead of "is"."""
+
+    @classmethod
+    @abstractmethod
+    def match(
+        cls: ConstraintT, node: NameNodes, expr: nodes.NodeNG, negate: bool = False
+    ) -> Optional[ConstraintT]:
+        """Return a new constraint for node matched from expr, if expr matches
+        the constraint pattern.
+
+        If negate is True, negate the constraint.
+        """
+
+    @abstractmethod
+    def satisfied_by(
+        self, inferred: Union[nodes.NodeNG, Type[util.Uninferable]]
+    ) -> bool:
+        """Return True if this constraint is satisfied by the given inferred value."""
+
+
+class NoneConstraint(Constraint):
+    """Represents an "is None" or "is not None" constraint."""
+
+    @classmethod
+    def match(
+        cls: ConstraintT, node: NameNodes, expr: nodes.NodeNG, negate: bool = False
+    ) -> Optional[ConstraintT]:
+        """Return a new constraint for node matched from expr, if expr matches
+        the constraint pattern.
+
+        Negate the constraint based on the value of negate.
+        """
+        if isinstance(expr, nodes.Compare) and len(expr.ops) == 1:
+            left = expr.left
+            op, right = expr.ops[0]
+            const_none = nodes.Const(None)
+            if op in {"is", "is not"} and (
+                matches(left, node)
+                and matches(right, const_none)
+                or matches(left, const_none)
+                and matches(right, node)
+            ):
+                negate = (op == "is" and negate) or (op == "is not" and not negate)
+                return cls(node=node, negate=negate)
+
+        return None
+
+    def satisfied_by(
+        self, inferred: Union[nodes.NodeNG, Type[util.Uninferable]]
+    ) -> bool:
+        """Return True if this constraint is satisfied by the given inferred value."""
+        # Assume true if uninferable
+        if inferred is util.Uninferable:
+            return True
+
+        if self.negate and matches(inferred, nodes.Const(None)):
+            return False
+        if not self.negate and not matches(inferred, nodes.Const(None)):
+            return False
+
+        return True
```

Comment:
```suggestion
        return bool(self.negate) ^ matches(inferred, nodes.Const(None))
```

```suggestion
        return bool(self.negate) ^ bool(matches(inferred, nodes.Const(None)))
```

I guess a xor would be slightly more efficient with a comparison less ? Not sure if we need the cast to bool on ``matches``.

Reply:
Thanks @Pierre-Sassoulas, I made a commit to use XOR. `self.negate` is a bool, and `matches` returns a `bool`, so we didn't need the explicit casting.